hudi-client/src/main/java/org/apache/hudi/client/HoodieWriteClient.java

```diff
@@ -487,6 +492,58 @@ private void saveWorkloadProfileMetadataToInflight(WorkloadProfile profile, Hood
     return updateIndexAndCommitIfNeeded(writeStatusRDD, hoodieTable, instantTime);
   }
 
+  /**
+   * Ensure that the current writerSchema is compatible with the latest schema of this dataset.
+   *
+   * When inserting/updating data, we read records using the schema saved in the data/log files
+   * and convert them to the GenericRecords with writerSchema. Hence, we need to ensure that
+   * this conversion can take place without errors.
+   *
+   * @param hoodieTable The Hoodie Table
+   * @param isUpsert If this is a check during upserts
+   * @throws HoodieUpsertException If schema check fails during upserts
+   * @throws HoodieInsertException If schema check fails during inserts
+   */
+  private void validateSchema(HoodieTable<T> hoodieTable, final boolean isUpsert)
```

Comment:
there is a pending PR to move this code into the new `table.action` package.. FYI.. this may need to move there as well